File: src/main/java/net/zenxarch/bot/defense/modules/KillAura.java
package net.zenxarch.bot.defense.modules;

import static net.zenxarch.bot.defense.DefenseStateManager.*;
import static net.zenxarch.bot.util.ClientPlayerHelper.*;

import net.minecraft.client.network.AbstractClientPlayerEntity;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.entity.mob.MobEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.item.MiningToolItem;
import net.minecraft.item.SwordItem;

public final class KillAura extends Module {

  public KillAura() { super("KillAura"); }

  @Override
  public void handleHostile(MobEntity me) {
    if (handleCrit())
      hitLiving(me);
  }

  @Override
  public void handlePlayer(AbstractClientPlayerEntity pe) {
    if (handleCrit())
      hitLiving(pe);
  }

  @Override
  public void handlePassive(MobEntity me) {
    if (handleCrit())
      hitLiving(me);
  }

  private void hitLiving(LivingEntity le) {
    performAction(() -> {
      if (!lookingAt(le))
        lookAt(le);
      pickItemSlot(findBestWeapon(le));
      hitEntity(le);
      return true;
    });
  }

  private int findBestWeapon(LivingEntity target) {
    var bestSlot = -1;
    var bestDamage = 0.0f;
    var inv = mc.player.getInventory();
    for (int i = 0; i < inv.main.size(); i++) {
      var dmg = getAttackDamage(inv.main.get(i), target);
      if (dmg > bestDamage) {
        bestDamage = dmg;
        bestSlot = i;
      }
    }

    if (getAttackDamage(inv.offHand.get(0), target) > bestDamage) {
      bestSlot = inv.main.size();
    }
    return bestSlot;
  }

  private float getAttackDamage(ItemStack stack, LivingEntity target) {
    if (stack.getItem() instanceof SwordItem || stack.getItem() instanceof
                                                    MiningToolItem) {
      return EnchantmentHelper.getAttackDamage(stack, target.getGroup());
    }
    return 0.0f;
  }

  private boolean handleCrit() {
    var canCrit = !(mc.player.isTouchingWater() || mc.player.isClimbing() ||
                    mc.player.isInLava() ||
                    mc.player.hasStatusEffect(StatusEffects.BLINDNESS) ||
                    mc.player.hasVehicle());
    var remainingTicks = getRemainingAttackCooldownTicks();
    if (canCrit) {
      if (mc.player.isOnGround()) {
        if (remainingTicks < 5)
          performAction(() -> {
            mc.player.jump();
            return true;
          });
        return false;
      } else
        return mc.player.fallDistance > 0;
    }
    return remainingTicks <= 0;
  }
}

File: src/main/scala/net/zenxarch/bot/defense/modules/KillAura.scala
package net.zenxarch.bot.defense.modules

import net.zenxarch.bot.defense.DefenseStateManager._
import net.zenxarch.bot.util.ClientPlayerHelper._

import net.minecraft.client.network.AbstractClientPlayerEntity
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.entity.mob.MobEntity
import net.minecraft.item.ItemStack

import net.minecraft.item.MiningToolItem
import net.minecraft.item.SwordItem
import net.zenxarch.bot.util.BaritoneUtils
import net.zenxarch.bot.util.WeaponUtils

class KillAura extends Module("KillAura"):
  import Module.mc

  override def handleHostile(me: MobEntity): Unit =
    BaritoneUtils.pausePathing()
    handleLiving(me)

  override def handlePlayer(pe: AbstractClientPlayerEntity): Unit =
    BaritoneUtils.pausePathing()
    handleLiving(pe)

  override def handlePassive(me: MobEntity): Unit =
    BaritoneUtils.pausePathing()
    handleLiving(me)

  override def handleNone(): Unit =
    BaritoneUtils.resumePathing()

  private def handleLiving(le: LivingEntity): Unit =
    performAction(() => internalHandleLiving(le))

  private def internalHandleLiving(le: LivingEntity): Boolean =
    if !lookingAt(le) then
      lookAt(le)
      syncRotation()
    val bestWeapon = findBestWeapon(le)

    // switch weapon
    if bestWeapon != mc.player.getInventory().selectedSlot then
      pickItemSlot(bestWeapon)
      return true

    val remainingTicks = getRemainingAttackCooldownTicks()
    val canCrit = !(mc.player.isTouchingWater() || mc.player.isClimbing() ||
      mc.player.isInLava() ||
      mc.player.hasStatusEffect(StatusEffects.BLINDNESS) ||
      mc.player.hasVehicle())

    if canCrit && mc.player.isOnGround then
      return if remainingTicks < 5 then
        doJump()
        true
      else false

    val shouldHit =
      remainingTicks < 1 && (if canCrit then mc.player.fallDistance > 0
                             else true)

    // hit the target
    if shouldHit then
      hitEntity(le)
      return true
    false

  private def findBestWeapon(target: LivingEntity): Int =
    return findBestInInventory(
      WeaponUtils.getAttackDamagePerSec(_, target.getGroup()),
      WeaponUtils.getBaseAttackDamagePerSec()
    )

  private def doJump() =
    val wasSneaking = mc.player.input.sneaking
    mc.player.input.sneaking = false
    mc.player.jump()
    mc.player.input.sneaking = wasSneaking
